Detector refactoring

c4ts0up opened the pull request:
# Description

Implementation of the Clustering and Centroid Analysis Universal Detector (CCA-UD) to detect poisoning backdoor attacks in deep neural networks (DNNs) described by [Guo, et al. (2021)](https://arxiv.org/abs/2301.04554).

Additionally, I decoupled the `ClusteringAnalyzer` as I found the OOP approach redundant. Finally, created a `PerformanceMonitor` utility to track hardware usage during the MNIST example created for CCA-UD.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Added unit tests for different components of CCA-UD. Used experimental results obtained by original authors to evaluate full model performance was on a par with proposed model (see notebook).

- [X] Unit tests (96% coverage)

**Test Configuration**:
- OS: Ubuntu 22.04.5 LTS (WSL2 on Windows 11 Home 24H2)
- Python version: 3.10.15
- ART version or commit number: 1.19.1
- TensorFlow: 2.14.0
- Keras: 2.14.0
- PyTorch: Not used
- MXNet version: 1.8.0.post0
- CUDA: V11.8.89

**Hardware**
- CPU: Ryzen 7 8845HS
- RAM: 16 GB
- GPU: NVIDIA GeForce RTX 4060 Laptop GPU

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] My changes have been tested using both CPU and GPU devices

c4ts0up commented:
Hi @beat-buesser ! Thanks for checking my PR 😄I actually wanted to ask you a couple of things:
1. I created the `art/performance_monitor.py` to easily visualize hardware usage in my example notebook. I couldn't find a similar utility in the framework or a third-party library (didn't know of one). Do you know of any utility for this? If not, should I keep the `art/performance_monitor.py` in this PR, or should I create a separate one for it?
2. In `art/defences/poison/clustering_centroid_analysis.py:408` I flattenned the features so that CCA-UD would work with other layers in the DNN. Pros: It is more flexible, allowing an ART user to choose any DNN layer for the feature extraction. Cons: it goes against the authors recommendation in the original paper. What do you think? Should I keep it, remove it or modify it? 
Thanks!

c4ts0up commented:
Rebased onto branch `dev_1.20.0`, fixed the merge conflicts and addressed the CodeQL warnings above. Ready for a new workflow run, @beat-buesser

beat-buesser commented on requirements_test.txt:
```diff
@@ -21,11 +21,13 @@ opencv-python
 
 tensorflow==2.18.1
 keras==3.10.0
+tensorflow-addons>=0.13.0
```
Can we avoid `tensorflow-addons` and replace it with `tensorflow`? We have included refactoring for 1.20 to remove it from ART by replacing all functions from `tensorflow-addons`. It started to affect upgrading to `tensorflow>=2.16.0` versions.

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2018
```
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2025
```

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+import numpy as np
+import pytest
+import tensorflow as tf
+from tensorflow import keras
```
```suggestion
```

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+        return self.cluster_labels_to_return
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1",
+    "keras",
+    "kerastf",
+    "pytorch",
+    "mxnet",
+    "non_dl_frameworks",
+)
```
We are removing TF1 and MXNet in ART 1.20.
```suggestion
@pytest.mark.skip_framework(
    "keras",
    "kerastf",
    "pytorch",
    "non_dl_frameworks",
)
```

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+        self.assertEqual({"A": 0, "B": 1, "C": 2, "D": 3}, reverse_mapping)
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1",
+    "keras",
+    "kerastf",
+    "pytorch",
+    "mxnet",
+    "non_dl_frameworks",
+)
```
```suggestion
@pytest.mark.skip_framework(
    "keras",
    "kerastf",
    "pytorch",
    "non_dl_frameworks",
)
```

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+        self.assertEqual(len(cluster_class_mapping), 0)
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1",
+    "keras",
+    "kerastf",
+    "pytorch",
+    "mxnet",
+    "non_dl_frameworkslearn",
+)
```
```suggestion
@pytest.mark.skip_framework(
    "keras",
    "kerastf",
    "pytorch",
    "non_dl_frameworkslearn",
)
```

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+        self.assertIsInstance(result, np.ndarray)
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1",
+    "keras",
+    "kerastf",
+    "pytorch",
+    "mxnet",
+    "non_dl_frameworkslearn",
+)
```
```suggestion
@pytest.mark.skip_framework(
    "keras",
    "kerastf",
    "pytorch",
    "non_dl_frameworkslearn",
)
```

beat-buesser commented on tests/defences/detector/poison/test_clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,1442 @@
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1",
+    "keras",
+    "kerastf",
+    "pytorch",
+    "mxnet",
+    "non_dl_frameworkslearn",
+)
```
```suggestion
    "keras",
    "kerastf",
    "pytorch",
    "non_dl_frameworkslearn",
)
```

beat-buesser commented on requirements_test.txt:
```diff
@@ -60,6 +62,8 @@ types-setuptools==80.9.0.20250529
 
 # other
 requests~=2.32.4
+psutil==7.0.0
+nvidia-ml-py==12.575.51
```
```suggestion
```

beat-buesser commented on art/attacks/poisoning/perturbations/__init__.py:
```diff
@@ -2,7 +2,7 @@
 """
 
-from art.attacks.poisoning.perturbations.image_perturbations import (
+from .image_perturbations import (
```
Why did you change this?

c4ts0up replied:
As this is the `__init__.py` of the homonymous package, I found it clearer to import locally: concise, decoupled and more standard.

beat-buesser commented on Dockerfile:
```diff
@@ -9,7 +9,7 @@ RUN pip3 install tensorflow==2.9.1 keras==2.9.0 numpy==1.22.4 scipy==1.8.1 matpl
                  torch==1.12.0 torchaudio==0.12.0 torchvision==0.13.0 catboost==1.0.6 GPy==1.10.0 \
                  lightgbm==3.3.2 xgboost==1.6.1 kornia==0.6.6 lief==0.12.1 pytest==7.1.2 pytest-pep8==1.0.6 \
-                 pytest-mock==3.8.2 requests==2.28.1
+                 pytest-mock==3.8.2 requests==2.28.1 umap-learn==0.5.7 psutil==7.0.0 nvidia-ml-py==12.575.51
```
```suggestion
                 pytest-mock==3.8.2 requests==2.28.1 umap-learn==0.5.7
```

beat-buesser commented on art/defences/detector/poison/clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,555 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2018
```
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2025
```

beat-buesser commented on art/defences/detector/poison/clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,555 @@
+import tensorflow as tf
+from sklearn.base import ClusterMixin
+from sklearn.cluster import DBSCAN
+from tensorflow.keras import Model, Sequential
```
Move import inside of the class to keep ART framework independent.

beat-buesser commented on art/defences/detector/poison/clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,555 @@
+from typing import TYPE_CHECKING, Any
+
+import numpy as np
+import tensorflow as tf
```
Move import inside of the class to keep ART framework independent.

beat-buesser commented on art/defences/detector/poison/clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,555 @@
+from sklearn.base import ClusterMixin
+from sklearn.cluster import DBSCAN
+from tensorflow.keras import Model, Sequential
+from umap import UMAP
```
Move import inside of the class to keep ART framework independent.

beat-buesser commented on art/defences/detector/poison/clustering_centroid_analysis.py:
```diff
@@ -0,0 +1,555 @@
+    return class_cluster_labels, cluster_class_mapping
+
+
+class ClusteringCentroidAnalysis(PoisonFilteringDefence):
```
Change name because it is a framework specific implementation.
```suggestion
class ClusteringCentroidAnalysisTensorFlowV2(PoisonFilteringDefence):
```

beat-buesser commented on run_tests.sh:
```diff
@@ -161,6 +164,7 @@ else
 
     declare -a art=("tests/test_data_generators.py" \
                     "tests/test_optimizers.py" \
+                    "tests/test_performance_monitor.py" \
```
```suggestion
```

beat-buesser commented on run_tests.sh:
```diff
@@ -25,6 +25,9 @@ then
     pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/defences/detector/poison/test_spectral_signature_defense.py --framework=$framework --durations=20 --durations-min=0
     if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/detector/poison/test_spectral_signature_defense.py tests"; fi
 
+    pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/defences/detector/poison/test_clustering_centroid_analysis.py --framework=$framework --durations=20 --durations-min=0
+    if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/detector/poison/test_clustering_centroid_analysis.py tests"; fi
+
```
```suggestion
```
Replace this with a new line after this line ""tests/defences/detector/poison/test_clustering_analyzer.py" \" around line 156.

beat-buesser commented on notebooks/poisoning_defense_clustering_centroid_analysis.ipynb:
Please update the file `/notebooks/READEME.md` accordingly by adding the new notebook following the alphabetical order.

beat-buesser commented on tests/test_performance_monitor.py:
Please remove this file.

beat-buesser commented on art/performance_monitor.py:
Please remove this file.

beat-buesser reviewed (changes requested):
Hi @c4ts0up Thank you for your pull request. I have completed a first review, we might need an additional round because of its size. Could you please implement the changes above and the following items:

- please add `@pytest.mark.framework_agnostic` to all tests that are framework independent, that way they will run only once
- test_clustering_centorid_analysis.py has 1442 lines of code, are they all necessary?
- thank you for sharing you performance monitor implementation, it looks really cool, but it is a bit outside of ART core mission and to what we can commit maintaining. Could you please remove it from this pull request?

c4ts0up commented:
> Hi @c4ts0up Thank you for your pull request. I have completed a first review, we might need an additional round because of its size. Could you please implement the changes above and the following items:
> 
> * please add `@pytest.mark.framework_agnostic` to all tests that are framework independent, that way they will run only once
> * test_clustering_centorid_analysis.py has 1442 lines of code, are they all necessary?
> * thank you for sharing you performance monitor implementation, it looks really cool, but it is a bit outside of ART core mission and to what we can commit maintaining. Could you please remove it from this pull request?

Hello @beat-buesser ! I've read your recommendations and applied them to the PR
1. Done! 
2. Agree that this is a lot of lines. I tried to test everything, including the private methods, as it made it easier for me to find bugs early on in the development; I had some trouble implementing the algorithm at the start, and I couldn't afford to run it whole to find in which step an error was happening (even non-crashing ones, that degraded model perfomance). Plus, had to add mocks and some additional setup for TF2. With a coverage of 90%, I think the length is reasonable. At least it keeps further developments on CCA-UD stable.
3. Completely agree. Removed!

I also added configurable batch processing to prevent OOM erros when running CCA-UD (specially in the example notebook).

Let me know if you'd recommend additional changes. Thanks again!

beat-buesser commented:
@c4ts0up Thank you, I'll take a look soon.

c4ts0up commented:
Got it! I also added some small, non-code fixes to the notebook's explanations. Thanks @beat-buesser !

beat-buesser reviewed (approved):
Hi @c4ts0up Thank you very much for your contribution to ART!
